### PR TITLE
[2015.8] Add __states__ to state modules, for cross-calling states

### LINKED ISSUE #21214: How to call other state in custom state code?
Hi, I'm writing a custom state in /srv/salt/_states/git_refresh.py, and I just want to call file.managed state in this py.

What should I do??
Thanks.

### PR DESCRIPTION
**Pull request by basepi**
Fixes #21214 

I thought it wasn't working, but it turned out to be my stupidity at work.

### CODE REVIEW

**Comment by thatch45** [comment]
@ryan-lane for the record it did not used to be this easy! This is only thanks to the new Lazy Loader we added in 2015.5

**Comment by basepi** [comment]
Valid!

**Comment by rvora** [comment]
Great. Can someone provide example code of how this is supposed to be used?

**Comment by basepi** [comment]
```
ret = __states__['file.managed'](name='/tmp/myfile', source='salt://myfile')
```

**Comment by jacksontj** [comment]
@basepi if this is staying in its current form (there are bugs, according to the original issue)-- please update the test case to check for it (was in my original patch for the same: https://github.com/jacksontj/salt/commit/54ee597a08f3370247de8d91b3886c6333933488)

**Comment by basepi** [comment]
Yep, I'm aware of the bugs, but it works for any state which doesn't utilize __running__ and the similar special variables. I'm trying to think of a solution for those.

**Comment by basepi** [comment]
And I'll add back in your test case when I start fixing those bugs.

**Comment by jacksontj** [comment]
Awesome :) And it sounds like we could use a test case for `__running__` ;) whenever we get a fix figured out

**Comment by basepi** [comment]
The problem with those variables is they are actually inserted at call time, by the state compiler. They're not packed globally.

Then again, I think Tom told me packing is done at runtime now, so it's possible we could change them to pack variables and make it work that way... that's a good path for investigation.

**Comment by jacksontj** [comment]
Hmm, I just finished played with that a bit and it gets a bit messier than I'd like... Basically we can make the pack include those by default-- and the load will inject them (I did it will all None's). Then we can replace the context manager in state.py to update the pack in the loader instead of the sys.modules. The problem is, the loader (at least today) does a setattr of pack-- which means once the module has been loaded it has its own copy of what pack's globals were at the time. This would mean if you kept the __state__ dict around for more than one run the modules loaded the first run would still have the stuff from the original run (mess!).

To make this work we'd need to make the loader pack in proxy objects instead-- which in some ways is good in general, but is yet more magic :/

**Comment by basepi** [comment]
@jacksontj I don't think this is technically an issue -- or, at least, I should clarify that it's no more an issue with those changes as without. Since we never call a state without putting new `__running__` (and friends) variables, the refresh happens at the beginning of a state run. Or am I missing something?

**Comment by jacksontj** [comment]
Yea, its a bit worse than that ;)

When the loader loads up all the state modules it creates a dict `__states__`. The loader injects all k,v pairs from another dict (`pack`) into the global namespeace of each and every module in `__states__` as it loads. The issue is, if you change `pack` (which a new state run would do) and `__states__` is being re-used, all modules loaded in the previous run will still have references to the original `pack`s globals-- not the new ones. This is because we do a setattr direct to the value, not to a proxy of some type.

Basically the loader is setting the globals to values instead of a pointer to `always be pack[key]`.

**Comment by basepi** [comment]
Ah, that makes sense. Thanks for looking into that, when I get a few minutes I'll play around with it myself, maybe we will have to go the proxy route.

I may also just fix the erroring file.managed, that will get rid of a lot of the problems. That's a popular candidate for cross-calling, and if I just fix it to only reference `__instance_id__` if it exists it will work with the current system. (It just won't support accumulators)

**Comment by jacksontj** [comment]
Yea, although TBH I wonder if we should pull this feature from 2015.8 (or at least config wrap it) since its going to have some `gotchas`, and we are supposed to be close to release.

**Comment by basepi** [comment]
Well, since enabling it doesn't *do* anything by default, config wrapping it is fairly pointless. People have to implement it in order to use it. So I think I'll just put a note in the release notes saying "this is experimental, certain states may not work until we work out the kinks"

**Comment by jacksontj** [comment]
Totally forgot-- yea they'll only run into issues if the call stuff in
__states__ :) Sounds good.
